middleware/metrics.go

```diff
@@ -0,0 +1,51 @@
+package middleware
+
+import (
+	"context"
+	"net/http"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+var metricsInfoKey = struct{}{}
+
+var (
+	RequestDurationOpts = prometheus.HistogramOpts{
+		Name:    "request_duration_seconds",
+		Help:    "Time (in seconds) spent serving this request.",
+		Buckets: prometheus.DefBuckets,
+	}
+)
+
+// MetricsInfo holds all of the metrics collected for a request.
+type MetricsInfo struct {
+	RequestDuration prometheus.Histogram
+	Start           time.Time
+}
+
+// GetMetrics retuns the metrics for this current context.
+func GetRequestMetrics(ctx context.Context) MetricsInfo {
+	metrics, ok := ctx.Value(metricsInfoKey).(*MetricsInfo)
+	if !ok || metrics == nil {
+		return MetricsInfo{
+			RequestDuration: nil,
+			Start:           time.Time{},
+		}
+	}
+	duration := time.Since(metrics.Start)
+	metrics.RequestDuration.Observe(duration.Seconds())
+	return *metrics
+}
+
+// NewMetricsMiddleware creates a middleware which produces metrics about a request, and tags the context with them.
+// Metrics info can be retrieved with `GetMetrics(ctx)`.
+func NewRequestMetricsMiddleware(next http.Handler) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, r *http.Request) {
+		ctx := context.WithValue(r.Context(), metricsInfoKey, &MetricsInfo{
```

Comment:
Is it useful to have this on the context?

Reply:
I think it depends on if we end up passing downstream, and pass a context down and back up. Something like a request that ends up calling a database, and passing the request metrics middleware all the way down - but reviewing that assumption, it's not clear that gives us more value than having downstream components basically time themselves (timestamping before, observing after) and report just that.